packages/react/src/use-cerbos-request.ts

```diff
@@ -0,0 +1,238 @@
+import type {
+  CheckResourceRequest,
+  CheckResourcesRequest,
+  CheckResourcesResponse,
+  CheckResourcesResult,
+  ClientWithPrincipal,
+  IsAllowedRequest,
+  RequestOptions,
+} from "@cerbos/core";
+import { useCallback, useEffect, useState } from "react";
+import { useDeepCompareMemoize } from "use-deep-compare-effect";
+
+import { useCerbos } from "./use-cerbos";
+
+/**
+ * The result of calling an async method on a client.
+ *
+ * @public
+ */
+export type AsyncResult<T> =
+  | { isLoading: true; data: undefined; error: undefined }
+  | { isLoading: false; data: T; error: undefined }
+  | { isLoading: false; data: undefined; error: Error };
+
+type Methods = "checkResource" | "checkResources" | "isAllowed";
+
+type Result<Method extends Methods> = Awaited<
+  ReturnType<ClientWithPrincipal[Method]>
+>;
+
+function useCerbosRequest<Method extends Methods>(
+  method: Method,
+  request: Parameters<ClientWithPrincipal[Method]>[0],
+  options?: Omit<RequestOptions, "signal">,
+): AsyncResult<Result<Method>> {
+  const [isLoading, setIsLoading] = useState(true);
+  const [data, setData] = useState<Result<Method>>();
+  const [error, setError] = useState<Error>();
+
+  const client = useCerbos();
+  const requestMemo = useDeepCompareMemoize(request);
+  const optionsMemo = useDeepCompareMemoize(options);
+
+  const load = useCallback<(signal: AbortSignal) => Promise<Result<Method>>>(
+    async (signal: AbortSignal): Promise<Result<Method>> =>
+      // @ts-expect-error -- https://github.com/microsoft/TypeScript/issues/30581
+      await client[method](requestMemo, {
+        ...optionsMemo,
+        signal,
+      }),
+    [client, method, optionsMemo, requestMemo],
+  );
+
+  useEffect(() => {
+    setIsLoading(true);
+    setData(undefined);
+    setError(undefined);
+
+    const abortController = new AbortController();
+
+    load(abortController.signal)
+      .then((data) => {
+        setIsLoading(false);
```

Comment:
I think the `then` gets enqueued after the promise gets resolved, so in the pathological case there might be a possibility that an abort happens in between the `fetch` resolving and this being called?

To be defensive, perhaps we should keep the `if (!aborted)` logic we had before.